### PR TITLE
Add tutorial on adding a new Wasm opcode

### PR DESCRIPTION
**Pull request by ngzhian**
@mathiasbynens ptal, content is exactly the same as the draft codelab, formatting changes and removing unnecessary prologue.

### CODE REVIEW

File: src/docs/webassembly-opcode.md [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+
+WebAssembly is implement in v8 in 3 ways:
+
+1. interpreter
+1. baseline compiler (Liftoff)
+1. optimizing compiler (Turbofan)
+
+The rest of this document focuses on the Turbofan pipeline, walking through how to add a new Wasm instruction and implement it in Turbofan.
+
+At a high level, Wasm instructions are compiled into a Turbofan graph, and we rely on the Turbofan pipeline to compile the graph into (ultimately) machine code. For more on Turbofan, check out the [v8 docs](https://v8.dev/docs/turbofan).
```
Comment (mathiasbynens) [inline src/docs/webassembly-opcode.md]: nit: TurboFan (multiple occurrences throughout the document)

File: src/docs/webassembly-opcode.md [hunk 2]
```diff
@@ -0,0 +1,430 @@
+---
+title: 'WebAssembly - adding a new opcode'
+---
+
+[WebAssembly](https://webassembly.org) (Wasm) is a binary instruction format for a stack-based virtual machine. This tutorial walks the reader through implementing a new WebAssembly instruction in v8.
+
+WebAssembly is implement in v8 in 3 ways:
```
Comment (mathiasbynens) [inline src/docs/webassembly-opcode.md]: implemented

Comment (mathiasbynens) [reply src/docs/webassembly-opcode.md]: nit: `s/v8/V8/` (multiple occurrences throughout the document)

Comment (RReverser) [reply src/docs/webassembly-opcode.md]: @mathiasbynens Note you can use Github suggestions for these sort of fixes, should make it a bit easier:

```suggestion
WebAssembly is implemented in V8 in 3 ways:
```

File: src/docs/webassembly-opcode.md [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+
+1. interpreter
```
Comment (RReverser) [inline src/docs/webassembly-opcode.md]: Nit: should these be capitalised?

File: src/docs/webassembly-opcode.md [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+
+Tip: finding the test name can be tricky, since the test definition is behind a macro. Use [Code Search](https://cs.chromium.org) to click around to discover the macro definitions.
+
+This error indicates that the compiler does not know of our new instruction - that will change in the next section.
+
+## Compiling Wasm into Turbofan
+
+In the introduction, we mentioned that Wasm instructions are compiled into a Turbofan graph, `wasm-compiler.cc` is the where this happens. Let's take a look at an example opcode, [I32Eqz](https://cs.chromium.org/chromium/src/v8/src/compiler/wasm-compiler.cc?l=716&rcl=686b68edf9f42c201c2b25bca9f4bef72ff41c0b):
+
+```
```
Comment (RReverser) [inline src/docs/webassembly-opcode.md]: Probably worth adding a language to the block? (not 100% sure how this Markdown is rendered, but usually you need it for the syntax highlight to work)

File: src/docs/webassembly-opcode.md [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+     case kArchCallCodeObject: {
+       if (HasImmediateInput(instr, 0)) {
+         Handle<Code> code = i.InputCode(0);
+```
+
+And this makes the test pass:
+
+Luckily for us `addl` is already implement. If our new opcode required writing a new assembly instruction implementation, we would add it to [`src/compiler/backend/x64/assembler-x64.cc`](https://cs.chromium.org/chromium/src/v8/src/codegen/x64/assembler-x64.cc), where the assembly instruction is encoded into bytes and emitted.
+
+Tip: to inspect the generated code, we can pass `--print-code` to `cctest`
```
Comment (RReverser) [inline src/docs/webassembly-opcode.md]:
```suggestion
Tip: to inspect the generated code, we can pass `--print-code` to `cctest`.
```

File: src/docs/webassembly-opcode.md [hunk 6]
```diff
@@ -0,0 +1,430 @@
+---
+title: 'WebAssembly - adding a new opcode'
+---
+
+[WebAssembly](https://webassembly.org) (Wasm) is a binary instruction format for a stack-based virtual machine. This tutorial walks the reader through implementing a new WebAssembly instruction in v8.
```
Comment (RReverser) [inline src/docs/webassembly-opcode.md]:
```suggestion
[WebAssembly](https://webassembly.org) (Wasm) is a binary instruction format for a stack-based virtual machine. This tutorial walks the reader through implementing a new WebAssembly instruction in V8.
```

File: src/docs/webassembly-opcode.md [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+1. baseline compiler (Liftoff)
+1. optimizing compiler (Turbofan)
+
+The rest of this document focuses on the Turbofan pipeline, walking through how to add a new Wasm instruction and implement it in Turbofan.
+
+At a high level, Wasm instructions are compiled into a Turbofan graph, and we rely on the Turbofan pipeline to compile the graph into (ultimately) machine code. For more on Turbofan, check out the [v8 docs](https://v8.dev/docs/turbofan).
+
+## Opcodes/Instructions
+
+We will define a new instruction that adds 1 to an [int32](https://webassembly.github.io/spec/core/syntax/types.html#syntax-valtype) (on the top of the stack).
```
Comment (RReverser) [inline src/docs/webassembly-opcode.md]: I'd mark these as code too:

```suggestion
We will define a new instruction that adds `1` to an [`int32`](https://webassembly.github.io/spec/core/syntax/types.html#syntax-valtype) (on the top of the stack).
```

File: src/docs/webassembly-opcode.md [hunk 8]
```diff
@@ -0,0 +1,430 @@
+---
+title: 'WebAssembly - adding a new opcode'
+---
+
+[WebAssembly](https://webassembly.org) (Wasm) is a binary instruction format for a stack-based virtual machine. This tutorial walks the reader through implementing a new WebAssembly instruction in V8.
+
+WebAssembly is implemented in V8 in 3 ways:
+
+1. Interpreter
```
Comment (RReverser) [inline src/docs/webassembly-opcode.md]: Nit: since these are capitalised, they should have dots in the end now.

**Comment by mathiasbynens** [comment]
Merging this as @RReverser is busy this week :)

**Pull request merged**